src/hotspot/share/opto/vector.cpp

```diff
@@ -311,39 +313,64 @@ void PhaseVector::expand_vbox_node(VectorBoxNode* vec_box) {
 Node* PhaseVector::expand_vbox_node_helper(Node* vbox,
                                            Node* vect,
                                            const TypeInstPtr* box_type,
-                                           const TypeVect* vect_type) {
+                                           const TypeVect* vect_type,
+                                           VectorSet &visited) {
+  // JDK-8304948 shows an example that there may be a cycle in the graph.
+  if (visited.test_set(vbox->_idx)) {
+    assert(vbox->is_Phi(), "should be phi");
+    return vbox; // already visited
+  }
+
+  // Handle the case when the allocation input to VectorBoxNode is a Proj.
+  // This is the normal case before expanding.
+  if (vbox->is_Proj() && vbox->in(0)->Opcode() == Op_VectorBoxAllocate) {
+    VectorBoxAllocateNode* vbox_alloc = static_cast<VectorBoxAllocateNode*>(vbox->in(0));
+    return expand_vbox_alloc_node(vbox_alloc, vect, box_type, vect_type);
+  }
+
+  // Handle the case when both the allocation input and vector input to
+  // VectorBoxNode are Phi. This case is generated after the transformation of
+  // Phi: Phi (VectorBox1 VectorBox2) => VectorBox (Phi1 Phi2).
+  // With this optimization, the relative two allocation inputs of VectorBox1 and
+  // VectorBox2 are gathered into Phi1 now. Similarly, the original vector
+  // inputs of two VectorBox nodes are in Phi2.
+  //
+  // See PhiNode::merge_through_phi in cfg.cpp for more details.
   if (vbox->is_Phi() && vect->is_Phi()) {
     assert(vbox->as_Phi()->region() == vect->as_Phi()->region(), "");
-    Node* new_phi = new PhiNode(vbox->as_Phi()->region(), box_type);
     for (uint i = 1; i < vbox->req(); i++) {
-      Node* new_box = expand_vbox_node_helper(vbox->in(i), vect->in(i), box_type, vect_type);
-      new_phi->set_req(i, new_box);
+      Node* new_box = expand_vbox_node_helper(vbox->in(i), vect->in(i),
+                                              box_type, vect_type, visited);
+      if (!new_box->is_Phi()) {
+        C->initial_gvn()->hash_delete(vbox);
```

Comment:
May I ask why is this needed? Thanks.

Reply:
> May I ask why is this needed? Thanks.

It is a constraint that nodes must be removed from hash table before modifying their inputs.  https://github.com/openjdk/jdk/blob/f4630cdc6f49812d147b9ba8a4ea4009968f0db2/src/hotspot/share/opto/node.hpp#L434

Reply:
Ah yes silly me, thanks a lot for your answers, I have no more questions.